### PR TITLE
Metrics

### LINKED ISSUE #1096: Create Prometheus Server & Grafana Stack
**Task summary**

Create a container stack that manages a Prometheus server and serves the Grafana webapp

**Specification reference**

The Prometheus server should collect metrics from the Forest process. The Grafana webapp should preload dashboards and come with documentation to update the dashboards.

**Other information and links**

### LINKED ISSUE #1097: Create Initial Set of Metrics on Forest
**Task summary**

Add first set of metrics to the Forest process and establish the patterns for defining, collecting, and updating metrics in Rust.

**Specification reference**

- Histogram to measure tipset processing time in ms
- Counter of GossipSub messages received by type
- Counter of tipsets received through GossipSub which fail validation
- Gauge of RocksDB size on disk
- Count of TipsetRangeSyncer future errors

**Other information and links**

### PR DESCRIPTION
**Pull request by olibearo**
**Summary of changes**

Changes introduced in this pull request:
- Adds the initial set of metrics for `forest` 
- Provides a simple way to set up the metrics stack (`prometheus` & `grafana`) using Docker

**Reference issue to close (if applicable)**

Closes #1096 & #1097  

**Other information and links**

### CODE REVIEW

**Comment by cryptoquick** [comment]
Nice! This looks like it measures block time and a few other things. How would you recommend I run it?

**Comment by cryptoquick** [comment]
Okay, and this works without prometheus running, correct?

**Comment by olibearo** [comment]
@cryptoquick that's correct

**Review by cryptoquick — CHANGES_REQUESTED** [review]
Got an error when trying to run this:
`
thread 'async-std/runtime' panicked at 'called `Result::unwrap()` on an `Err` value: InconsistentCardinality { expect: 7, got: 1 }', /home/hunter/.cargo/registry/src/github.com-1ecc6299db9ec823/prometheus-0.12.0/src/vec.rs:258:49
`
Screenshot from 2021-05-19 19-02-34

**Comment by cryptoquick** [comment]
Also, you're missing the following metric as defined in the issue ACs:
- Gauge of RocksDB size on disk

Maybe try measuring this after every flush:
https://doc.rust-lang.org/std/os/unix/fs/trait.MetadataExt.html#tymethod.size

**Comment by olibearo** [comment]
@cryptoquick I didn't realize that moving this from draft to open would auto request review, that's my bad. I still have much to do, including the items you mentioned.  I don't see a button that will move this from open -> draft again.

**Comment by cryptoquick** [comment]
Oh, gotcha! Be sure to let us know when it's RFR. Also LMK if you think I could help in some way!

File: .circleci/config.yml [hunk 1]
```diff
@@ -271,7 +271,7 @@ jobs:
       - run:
           name: Check for known security issues in dependencies
           # TODO remove ignore when https://github.com/filecoin-project/rust-fil-proofs/issues/1403 resolved
-          command: cargo audit --ignore RUSTSEC-2021-0011
+          command: cargo audit --ignore RUSTSEC-2021-0011 --ignore RUSTSEC-2021-0070
```
Comment (cryptoquick) [inline .circleci/config.yml:274]: Can you leave a comment like the one on 273?
Also, be sure to label that 273 is for `RUSTSEC-2021-0011`, while the new comment is for `RUSTSEC-2021-0070`, so we know which is which.

**Review by cryptoquick — CHANGES_REQUESTED** [review]

**Comment by ec2** [comment]
i know this isnt R4R yet, but you've incorrectly grouped Libp2p metrics (eg HELLO_REQUEST)  as Gossip metrics.

File: blockchain/chain_sync/src/chain_muxer.rs [hunk 2]
```diff
@@ -140,7 +146,7 @@ pub struct ChainMuxer<DB, TBeacon, V, M> {
     /// Will mark any invalid blocks and all childen as bad in this bounded cache
     bad_blocks: Arc<BadBlockCache>,
 
-    /// Incoming network events to be handled by syncer
+    ///   network events to be handled by syncer
```
Comment (cryptoquick) [inline blockchain/chain_sync/src/chain_muxer.rs:149]: Typo

**Review by cryptoquick — CHANGES_REQUESTED** [review]
Just found one obvious typo.

I'm able to compile this and it runs fine, but I'm not able to test it on my laptop because, although I have docker configured on my desktop, I'm reluctant to do similar on my laptop, so my feedback on this will be somewhat limited for now.

**Comment by ec2** [comment]
It LGTM. Im just setting up Docker on BigChungus right now to see this in live action

**Review by cryptoquick — APPROVED** [review]
The feedback I've provided has been addressed. @ec2, be sure to let us know how your Docker deploy works out.

**Review by ec2 — APPROVED** [review]
Just tested out with the new metrics. Looks perfect!

File: forest/src/cli/mod.rs [hunk 3]
```diff
@@ -107,6 +107,8 @@ pub struct DaemonOpts {
         help = "Amount of Peers we want to be connected to (default is 75)"
     )]
     pub target_peer_count: Option<u32>,
+    #[structopt(long, help = "Encrypt the keystore (default = true)")]
+    pub encrypt_keystore: Option<bool>,
```
Comment (creativcoder) [inline forest/src/cli/mod.rs:111]: are these changes part of this PR?

**Review by creativcoder — APPROVED** [review]
Looks great! The only thing I felt is if `Metrics` could be global so we wouldn't have to pass it around everywhere. Alternatively, it could be a trait with default methods, which structs could then implement.

I tried running docker compose on aws machine but it fails. Posted in detail on slack.

File: forest/src/cli/mod.rs (hunk 3 above)
Comment (olibearo) [reply forest/src/cli/mod.rs:111]: Yes

**Comment by olibearo** [comment]
> Looks great! The only thing I felt is if `Metrics` could be global so we wouldn't have to pass it around everywhere. Alternatively, it could be a trait with default methods, which structs could then implement.
> 
> I tried running docker compose on aws machine but it fails. Posted in detail on slack.

This was an issue with the `docker` installation on Rahul's machine. I've updated the documentation to make installation easier.

File: blockchain/chain_sync/src/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use lazy_static::lazy_static;
+use prometheus::{
+    core::{AtomicU64, GenericCounter, GenericCounterVec, GenericGauge, Opts},
+    Histogram, HistogramOpts,
+};
+
+lazy_static! {
```
Comment (creativcoder) [inline blockchain/chain_sync/src/metrics.rs:10]: Kudos for making this a singleton! One more suggestion, why don't we try once_cell: https://github.com/async-rs/async-std/issues/406 as people speak highly of it compared to lazy_static.

Comment (olibearo) [reply blockchain/chain_sync/src/metrics.rs:10]: Yeah, I looked into it, but the stdlib impl is only available in nightly and despite that, it's a container type that would require every routine that needs the `OnceCell<T>`  to call `get` to retrieve the inner `T`.

Comment (olibearo) [reply blockchain/chain_sync/src/metrics.rs:10]: I'm all ears if you understand the benefits we might get from adding that getter every time we collect a metric.

Comment (ec2) [reply blockchain/chain_sync/src/metrics.rs:10]: From my understanding, the benefits would be:
- Lack of macros
- Faster hot get
I am not sure if that is enough for us to switch over at this point for this.

Comment (olibearo) [reply blockchain/chain_sync/src/metrics.rs:10]: Why is it faster than accessing a global static variable?

Is the concern with macros the incremental increase in compilation cost?

Comment (creativcoder) [reply blockchain/chain_sync/src/metrics.rs:10]: oh, you don't need to use the OnceCell container. Instead you can use: https://docs.rs/once_cell/1.8.0/once_cell/#lazy-initialized-global-data. Anyways if you feel the refactor is quite complex, then lazy_static it is! As for perf improvements, we will have to measure to see the benefits in context to forest's access patterns.

Comment (ec2) [reply blockchain/chain_sync/src/metrics.rs:10]: Not sure why its faster on hot get. That's just what someone said. But i dont think its significant. 
The concern with macros is increased compilation cost as well as the fact that there is code being generated. That being said, those are "meh" concerns for me... I'm ok with lazy_static. We use lazy_static in other parts of the codebase

Comment (olibearo) [reply blockchain/chain_sync/src/metrics.rs:10]: I don’t feel enticed, seems more fashion than function

**Pull request merged**